File: day2/part2.py
def isInvalid(number):
    # print(f"Checking number {number}")
    line = str(number)
    # print(f"Length of number is {len(line)}")
    # print(f"Checking lengths from 1 to {len(line)//2}")
    isInvalid = any(isInvalidForLength(number, length) for length in range(1, len(line)//2 + 1))
    if isInvalid: print(f"Number {number} is invalid")
    return isInvalid


def isInvalidForLength(number, length):
    # print(f"Checking number {number} for length {length}")
    line = str(number)
    if len(line) % length != 0:
        return False
    
    chunks = list(chunk_list(line, length))
    # print(f"Chunks: {chunks}")
    isInvalid = all(chunk == chunks[0] for chunk in chunks)
    if isInvalid: print(f"Number {number} is invalid for length {length}")
    return isInvalid






def chunk_list(lst, n):
    """Yield successive n-sized chunks from lst."""
    for i in range(0, len(lst), n):
        yield lst[i:i + n]




def getNumbersFromRanges(file):
    numbers = []
    ranges = file.split(",")
    for r in ranges:
        print(r)
        cleanline = r.strip()
        parts = r.split("-")
        start = int(parts[0])
        end = int(parts[1])
        # print(f"Generating numbers from {start} to {end}")
        # for i in range(start, end + 1):
        #     print(i)
        #     numbers.append(i)
        numbers.extend(range(start, end + 1))
    return numbers



# Open the file in read mode and use readlines()
with open("input.txt", "r") as f:
    file = f.read()
    numbers = getNumbersFromRanges(file)
    invalidNumbers = [number for number in numbers if isInvalid(number)]
    print(invalidNumbers)
    print(len(invalidNumbers))
    print(sum(invalidNumbers))

File: day2/node/part2.js
const fs = require('fs');
const path = require('path');

/**
 * Check if a number is invalid
 * A number is invalid if it can be divided into equal-length chunks that are all identical
 * @param {number} number - The number to check
 * @returns {boolean} - True if invalid, false if valid
 */
function isInvalid(number) {
  const line = String(number);
  const isInvalidFlag = Array.from({ length: Math.floor(line.length / 2) }, (_, i) => i + 1)
    .some(length => isInvalidForLength(number, length));
  
  if (isInvalidFlag) {
    console.log(`Number ${number} is invalid`);
  }
  return isInvalidFlag;
}

/**
 * Check if a number is invalid for a specific chunk length
 * @param {number} number - The number to check
 * @param {number} length - The chunk length
 * @returns {boolean} - True if invalid for this length, false otherwise
 */
function isInvalidForLength(number, length) {
  const line = String(number);
  if (line.length % length !== 0) {
    return false;
  }
  
  const chunks = chunkString(line, length);
  const isInvalidFlag = chunks.every(chunk => chunk === chunks[0]);
  
  if (isInvalidFlag) {
    console.log(`Number ${number} is invalid for length ${length}`);
  }
  return isInvalidFlag;
}

/**
 * Split a string into chunks of specified length
 * @param {string} str - The string to chunk
 * @param {number} n - The chunk size
 * @returns {string[]} - Array of chunks
 */
function chunkString(str, n) {
  const chunks = [];
  for (let i = 0; i < str.length; i += n) {
    chunks.push(str.substring(i, i + n));
  }
  return chunks;
}

/**
 * Parse ranges from file content and generate all numbers in those ranges
 * @param {string} fileContent - The file content containing comma-separated ranges
 * @returns {number[]} - Array of all numbers from all ranges
 */
function getNumbersFromRanges(fileContent) {
  const numbers = [];
  const ranges = fileContent.split(',');
  
  for (const r of ranges) {
    console.log(r);
    const cleanLine = r.trim();
    const parts = r.split('-');
    const start = parseInt(parts[0], 10);
    const end = parseInt(parts[1], 10);
    
    for (let i = start; i <= end; i++) {
      numbers.push(i);
    }
  }
  
  return numbers;
}

// Main execution
const filePath = path.join(__dirname, 'input.txt');
const fileContent = fs.readFileSync(filePath, 'utf8');
const numbers = getNumbersFromRanges(fileContent);
const invalidNumbers = numbers.filter(number => isInvalid(number));

console.log(invalidNumbers);
console.log(invalidNumbers.length);
console.log(invalidNumbers.reduce((sum, num) => sum + num, 0));